tools: normalized options for build-server/executable

lazan opened the pull request:
This patch: Normalizes common options that both build-server and build-executable uses.
This will help keeping the two from diverging.

Resolves #567

lazan committed:
tools: normalized options for build-server/executable

This patch: Normalizes common options that both build-server and
build-executable uses.
This will help keeping the two from diverging.

Resolves #567

lazan committed:
argument: improved --help to print only one level

Before: we used some _terminal_ output when we print the general help,
no for a specific option.
This is not consistent with specific option, and does not provide enough
information to the user.

Now: we use the regular print, but only print _on level deep_.

lazan committed:
tools: added documentation casual-make target

This patch:
* generates casual-build-server --help -> markdown
* generates casual-build-executable --help -> markdown

OlofGullnas commented:
I have started looking at this PR.

Factoring out the common part of the command options for the "build tools" is a good approach :-)

I tried to build the branch, but I get an error:

```
buildserver: /export/home/guo/c_1.8/casual-1.8/middleware/example/server/bin/casual-example-server
buildserver: /export/home/guo/c_1.8/casual-1.8/middleware/example/server/bin/casual-example-resource-server
error:
[casual:invalid_argument] failed to compile - capture: { exit: 1, standard: { out: , error: /usr/lib64/gcc/x86_64-suse-linux/13/../../../../x86_64-suse-linux/bin/ld: /export/home/guo/c_1.8/casual-1.8/middleware/common/bin/libcasual-mockup-rm.so: undefined reference to `casual::common::transaction::xid::null()'
/usr/lib64/gcc/x86_64-suse-linux/13/../../../../x86_64-suse-linux/bin/ld: /export/home/guo/c_1.8/casual-1.8/middleware/common/bin/libcasual-mockup-rm.so: undefined reference to `casual::common::transaction::Context::instance()'
/usr/lib64/gcc/x86_64-suse-linux/13/../../../../x86_64-suse-linux/bin/ld: /export/home/guo/c_1.8/casual-1.8/middleware/common/bin/libcasual-mockup-rm.so: undefined reference to `casual::common::unittest::log'
/usr/lib64/gcc/x86_64-suse-linux/13/../../../../x86_64-suse-linux/bin/ld: /export/home/guo/c_1.8/casual-1.8/middleware/common/bin/libcasual-mockup-rm.so: undefined reference to `casual::common::transaction::Context::resource_registration(casual::common::strong::Type<int, casual::common::strong::resource::policy>, xid_t*)'
collect2: error: ld returned 1 exit status
}}: invalid_argument

aborting due to errors
error building...
[infra] guo@bvin01-unisb-01:~/c_1.8/casual-1.8 
``` 
The build error is not obviously cobbected to the changes in the PR.

I will start a new build after a "casual-make --debug clean" in case this is caused by a inconsistent build. 

(Or is this caused by the build referencing an old "installed" verion of casual. or something like that. My last build of Casual was some time ago,)

OlofGullnas commented:
There was an old file common/bin/libcasual-mockup-rm.so in the build tree, When I removed it the build worked OK....

Will continue looking into the PR...

Unit test fails. There are more old files in the build tree...

OlofGullnas commented:
The failing unit tests use queues. I suspect my problem with the unit tests is my old issue of having an NFS mounted home directory. This causes sqlite databases to be on an NFS  mounted disk which is not good.

OlofGullnas commented:
I looked at the help for some build commands. I noticed a difference in --verbose. The build command for executable and server has this as an option without argument. casual-build-resource-proxy has this as an option with an argument (0/1/true/false...?)

There is also a difference in the option to keep intermediate file. For server and executable this is -k or --keep. For resource proxy it is -s or --keep-source. (-k is used for the resource key).

I also noticed that casual-build-resource-proxy complains with 

 `[casual:invalid_argument] resource-key:  not found: invalid_argument
`
after the help text when you do casual-build-resource-proxy --help. 

Backward compatibility is an obvious concern/issue when changing command options and may make it infeasible to use the new shared settings also for casual-build-resource-proxy.

OlofGullnas reviewed (approved):
Changes are reasonable.
I would have named things a little different. I would like to emphasize that some "settings" are command line options/settings that are common to several (potentially all) of the "build" commands. I.e. for servers, executables and resource prexies. Making the build of resource proxies more uniform with the other would break backward compatibility.

Some combination of the words "shared", "build" and {"settings" or "options"}. The word "common" should probably be avoided :-)

lazan commented:
> Some combination of the words "shared", "build" and {"settings" or "options"}. The word "common" should probably be avoided :-)

The _fqn_ is `tools::build::setting::Directive`. It's not _options_ per se, it's what argument::options sets/assign. I don't want to use the word _option(s)_.  Hmm, I can't come up with something that feels ok and better.

`tools::build::settings::Mandatory` it is.

lazan committed:
tools: build::setting::Directive -> build::setting::Mandatory

This patch changes the name from tools::build::setting::Directive to
tools::build::setting::Mandatory

lazan committed:
thirdparty: updated function2 single header

Previous version where using std::aligned_storage_t. Deprecated in c++23
